lib/Reader/readerir.cpp

```diff
@@ -945,6 +971,118 @@ void GenIR::createSafepointPoll() {
 
 bool GenIR::doTailCallOpt() { return JitContext->Options->DoTailCallOpt; }
 
+void GenIR::pushDebugScope(llvm::Function *F) {
+  LLILCDbgInfo.LexicalBlocks.push_back(LLILCDbgInfo.FnScopeMap[F]);
+}
+
+void GenIR::popDebugScope() {
+  LLILCDbgInfo.LexicalBlocks.pop_back();
+}
+
+// Set the Debug Location for the current instruction
+void GenIR::setDebugLocation(uint32_t CurrOffset, uint32_t IsCall) {
+  DIScope *Scope;
+  if (LLILCDbgInfo.LexicalBlocks.empty())
+    Scope = LLILCDbgInfo.TheCU;
+  else
+    Scope = LLILCDbgInfo.LexicalBlocks.back();
+  DebugLoc Loc = DebugLoc::get(CurrOffset, IsCall, Scope);
+
+  LLVMBuilder->SetCurrentDebugLocation(Loc);
+}
+
+llvm::DISubroutineType *GenIR::createFunctionType(llvm::Function *F,
+                                                  DIFile *Unit) {
+  SmallVector<Metadata *, 8> EltTys;
+  // for each param, etc. convert the type to DIType and add it to the array.
+  FunctionType *FunctionTy = F->getFunctionType();
+
+  unsigned i = 0;
+
+  DIType *ReturnTy = convertType(Function->getReturnType());
+  EltTys.push_back(ReturnTy);
+
+  for (auto ParamIterator = FunctionTy->param_begin();
+       ParamIterator != FunctionTy->param_end(); ParamIterator++) {
```

Comment:
Nit: lift the call to `param_end()` out of the loop or just use `for (Type *Ty : FunctionTy->params())`